common/solver/batch_gmres_kernels.hpp.inc

```diff
@@ -0,0 +1,985 @@
+    const gko::batch_dense::BatchEntry<const ValueType> &z_shared_entry,
+    const uint32 &converged)
+{
+    using real_type = typename gko::remove_complex<ValueType>;
+
+    const gko::batch_dense::BatchEntry<ValueType> v_naught_shared_entry{
+        V_shared_entry.values, V_shared_entry.stride, z_shared_entry.num_rows,
+        V_shared_entry.num_rhs};
+
+    __shared__
+        UninitializedArray<real_type, batch_config<ValueType>::max_num_rhs>
+            z_norms_sh;
+
+    const gko::batch_dense::BatchEntry<real_type> z_norms_shared_entry{
+        z_norms_sh, z_shared_entry.stride, 1, z_shared_entry.num_rhs};
+
+    compute_norm2(z_shared_entry, z_norms_shared_entry);
+
+    __syncthreads();
+
+    for (int li = threadIdx.x;
+         li < v_naught_shared_entry.num_rows * v_naught_shared_entry.num_rhs;
+         li += blockDim.x) {
+        int r = li / v_naught_shared_entry.num_rhs;
+        int c = li % v_naught_shared_entry.num_rhs;
+
+        const uint32 conv = converged & (1 << c);
+
+        if (conv) {
+            continue;
+        }
+
+
+        v_naught_shared_entry.values[r * v_naught_shared_entry.stride + c] =
+            z_shared_entry.values[r * z_shared_entry.stride + c] /
+            static_cast<ValueType>(z_norms_shared_entry.values[c]);
+    }
+
+
+    for (int li = threadIdx.x;
+         li < s_shared_entry.num_rows * s_shared_entry.num_rhs;
+         li += blockDim.x) {
+        int r = li / s_shared_entry.num_rhs;
+        int c = li % s_shared_entry.num_rhs;
+
+        const uint32 conv = converged & (1 << c);
+
+        if (conv) {
+            continue;
+        }
+
+        s_shared_entry.values[r * s_shared_entry.stride + c] =
+            zero<ValueType>();
+
+        if (r == 0) {
+            s_shared_entry.values[r * s_shared_entry.stride + c] =
+                z_norms_shared_entry.values[c];
+        }
+    }
+}
+
+
+template <typename BatchMatrixType_entry, typename ValueType, typename PrecType>
+__device__ __forceinline__ void arnoldi(
+    const BatchMatrixType_entry &A_global_entry, const int inner_iter,
+    const gko::batch_dense::BatchEntry<ValueType> &V_shared_entry,
+    const gko::batch_dense::BatchEntry<ValueType> &H_shared_entry,
+    const gko::batch_dense::BatchEntry<ValueType> &w_shared_entry,
+    const PrecType &prec_shared, const uint32 &converged)
+{
+    using real_type = typename gko::remove_complex<ValueType>;
+    const int &i = inner_iter;
+    const int num_rows = w_shared_entry.num_rows;
+    const int num_rhs = w_shared_entry.num_rhs;
+
+    const gko::batch_dense::BatchEntry<const ValueType> v_i_shared_entry{
+        &V_shared_entry.values[i * V_shared_entry.stride * num_rows],
+        V_shared_entry.stride, num_rows, V_shared_entry.num_rhs};
+
+    __shared__
+        UninitializedArray<ValueType, batch_config<ValueType>::max_num_rows *
+                                          batch_config<ValueType>::max_num_rhs>
+            w_temp_sh;
```

Comment:
I think you should allocate this in dynamic shared memory and pass it in. Something proportional to the number of rows might be too big to allocate statically without affecting occupancy.

Reply:
Done